builds/fabric8-che/plugins/keycloak-plugin-token-provider/src/main/java/com/redhat/che/keycloak/token/provider/util/KeycloakUserValidator.java

```diff
@@ -0,0 +1,236 @@
+import io.fabric8.openshift.client.DefaultOpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftConfig;
+import io.fabric8.openshift.client.OpenShiftConfigBuilder;
+
+/**
+ * Validates that a user, specified by a keycloak token, matches the user that
+ * owns the OpenShift namespace where this class is loaded.
+ *
+ * @author amisevsk
+ */
+@Singleton
+public final class KeycloakUserValidator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KeycloakUserValidator.class);
+    private static final String ACCESS_TOKEN = "access_token";
+    private static final int CACHE_TIMEOUT_MINUTES = 10;
+
+    private final KeycloakTokenProvider keycloakTokenProvider;
+
+    /**
+     * Cache that stores mappings from keycloak token to OpenShift token. Timeout is
+     * equal to CACHE_TIMEOUT_MINUTES.
+     *
+     * @see TokenLoader
+     */
+    private final LoadingCache<String, String> keycloakToOpenshiftTokenCache;
+
+    /**
+     * Cache that stores mappings from OpenShift tokens to Users. Timeout is equal to
+     * CACHE_TIMEOUT_MINUTES.
+     *
+     * @see UserLoader
+     */
+    private final LoadingCache<String, String> openShiftTokenToUserCache;
+
+    @Inject
+    public KeycloakUserValidator(KeycloakTokenProvider tokenProvider) {
+        this.keycloakTokenProvider = tokenProvider;
+
+        this.keycloakToOpenshiftTokenCache = CacheBuilder.newBuilder()
+                                                         .maximumSize(100)
+                                                         .expireAfterWrite(CACHE_TIMEOUT_MINUTES, TimeUnit.MINUTES)
+                                                         .build(new TokenLoader());
+        this.openShiftTokenToUserCache = CacheBuilder.newBuilder()
+                                                     .maximumSize(100)
+                                                     .expireAfterWrite(CACHE_TIMEOUT_MINUTES, TimeUnit.MINUTES)
+                                                     .build(new UserLoader());
+    }
+
+    /**
+     * Check if user specified by keycloak token matches the user associated with the current
+     * Che instance by making request against the relevant endpoints. To improve performance,
+     * API responses are cached for a period of time.
+     *
+     * @param keycloakToken the value of the auth header (i.e. the keycloak token)
+     * @return true if the user is authorized to access this instance of Che, false otherwise.
+     */
+    public boolean matchesUsername(String keycloakToken) {
+
+        if (isNullOrEmpty(keycloakToken)) {
+            LOG.info("Received request with null or empty auth value. Returning false");
+            return false;
+        }
+        
+        LOG.debug("Keycloak token = {}", keycloakToken);
+
+        String projectOwner = getOpenShiftProjectOwner();
+        if ("eclipse".equals(projectOwner)) {
```

Comment:
well we discussed this yesterday during the standup, and being able to test the other Keycloak-related features on Minishift seems still useful especially regarding remote debugging, that wouldn't be available on OSO.